### PR TITLE
pg_upgrade 5.0->5.0 take2^2

### PR DESCRIPTION
**Pull request by danielgustafsson**
This is a remix of #2646, rebased on master and with additional small fixes and a new testmode (since @hlinnaka is on relaxing on the beach I forked his branch). There are still pg_dump bugs that will cause the pg_upgrade test to fail, so in order to get this in without all that blocking a new smoketest mode is added. The smoketest upgrades only the QD, which will catch breakage in Oid synchronization due to new/altered objects. It is **Not** a test of a cluster upgrade, but a test that pg_upgrade is broken while the other blockers are being worked on. The full upgrade testmode can be invoked with `make -C contrib/pg_upgrade installcheck` and currently succeeds to upgrade the cluster but fails in diffing the old/new due to pg_dump issues.

As mentioned in #2646, getting this in puts pg_upgrade in a much improved state while we continue to work on the remaining issues (which are to a large degree external to pg_upgrade).

Ping @Eulerizeit

### CODE REVIEW

File: contrib/pg_upgrade/check.c [hunk 1]
```diff
@@ -933,3 +938,120 @@ check_fts_fault_strategy(migratorContext *ctx)
...
+	bool			found = false;
+	int				dbnum;
+
+	prep_status(ctx, "Checking for non-covering indexes on partitioned AO tables");
+
+	snprintf(output_path, sizeof(output_path), "%s/mismatched_aopartition_indexes.txt",
+			 ctx->cwd);
+
+	snprintf(query, sizeof(query),
+			 "SELECT DISTINCT ao.relid, inh.inhrelid "
```
Comment (andreasscherbaum) [inline contrib/pg_upgrade/check.c]: This only lists the OIDs, right? Any chance to select the index name in question as well? Would certainly make it easier for the person who has to drop the index.

File: contrib/pg_upgrade/dump.c [hunk 2]
```diff
@@ -28,6 +28,37 @@ generate_old_dump(migratorContext *ctx)
 	check_ok(ctx);
 }
 
+static char *
+get_preassigned_oids_for_db(migratorContext *ctx, char *line)
+{
+	char	   *dbname;
+	int			dbnum;
+
+	// FIXME: parsing the dump like this is madness.
```
Comment (andreasscherbaum) [inline contrib/pg_upgrade/dump.c]: Will this be fixed? ;-)

File: contrib/pg_upgrade/pg_upgrade.c [hunk 3]
```diff
@@ -472,6 +465,15 @@ create_new_objects(migratorContext *ctx)
 	 */
 	if (GET_MAJOR_VERSION(ctx->old.major_version) <= 802)
 		new_gpdb5_0_invalidate_indexes(ctx, ctx->check, CLUSTER_NEW);
+	else
+	{
+		/* TODO: Bitmap indexes are not supported, so mark them as invalid. */
```
Comment (andreasscherbaum) [inline contrib/pg_upgrade/pg_upgrade.c]: Is this still a todo?

File: doc/src/sgml/ref/create_type.sgml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <!--
-$PostgreSQL: pgsql/doc/src/sgml/ref/create_type.sgml,v 1.73.2.1 2008/05/27 18:05:20 tgl Exp $
+$PostgreSQL: pgsql/doc/src/sgml/ref/create_type.sgml,v 1.80 2009/12/26 16:55:21 momjian Exp $
 PostgreSQL documentation
```
Comment (andreasscherbaum) [inline doc/src/sgml/ref/create_type.sgml]: That's coming from upstream, I assume?

**Comment by andreasscherbaum** [comment]
Just one minor comment: looks like that in the comment sometimes it's "Oid" or "Oids" and sometimes it's "OID" and "OIDs".

File: doc/src/sgml/ref/create_type.sgml (hunk 4 above)
Comment (danielgustafsson) [inline doc/src/sgml/ref/create_type.sgml]: Correct.

File: contrib/pg_upgrade/dump.c (hunk 2 above)
Comment (danielgustafsson) [inline contrib/pg_upgrade/dump.c]: Patches welcome :wink: =)

File: contrib/pg_upgrade/check.c (hunk 1 above)
Comment (danielgustafsson) [inline contrib/pg_upgrade/check.c]: Thats a good suggestion, I will add that to the list of things we should do.

File: contrib/pg_upgrade/pg_upgrade.c (hunk 3 above)
Comment (danielgustafsson) [inline contrib/pg_upgrade/pg_upgrade.c]: It's a TODO, invalidating them is a fine for now.

**Comment by danielgustafsson** [comment]
Pushed to master.

**Pull request closed without merging**